x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/aggs/mapreduce/InternalMapReduceAggregation.java

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.ml.aggs.mapreduce;
+
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.common.util.BigArrays;
+import org.elasticsearch.search.aggregations.AggregationExecutionException;
+import org.elasticsearch.search.aggregations.AggregationReduceContext;
+import org.elasticsearch.search.aggregations.InternalAggregation;
+import org.elasticsearch.search.aggregations.support.SamplingContext;
+import org.elasticsearch.search.profile.SearchProfileResults;
+import org.elasticsearch.xcontent.XContentBuilder;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+public final class InternalMapReduceAggregation extends InternalAggregation {
+
+    private final AbstractMapReducer<?, ?, ?> mapReducer;
+    private Writeable mapReduceContext;
```

Comment:
Instead of making this a `Writeable`, could it be made a generic type that extends both `XContent` and `Writeable`?  That would avoid the casting concerns.

Reply:
Great, that sounds like the ideal solution.